### PR TITLE
[SPARK-9104][CORE] Expose Netty memory metrics in Spark

### PR DESCRIPTION
**Pull request by jerryshao**
## What changes were proposed in this pull request?

This PR exposes Netty memory usage for Spark's `TransportClientFactory` and `TransportServer`, including the details of each direct arena and heap arena metrics, as well as aggregated metrics. The purpose of adding the Netty metrics is to better know the memory usage of Netty in Spark shuffle, rpc and others network communications, and guide us to better configure the memory size of executors.

This PR doesn't expose these metrics to any sink, to leverage this feature, still requires to connect to either MetricsSystem or collect them back to Driver to display.

## How was this patch tested?

Add Unit test to verify it, also manually verified in real cluster.

### CODE REVIEW

**Comment by jerryshao** [comment]
@jsoltren @squito I take the liberty to get another shot of exposing Netty memory metrics. This PR doesn't address any display related things, only expose memory usage from Netty client factory and server. The following PRs could leverage this to display shuffle, RPC as well as other component's memory usage for Netty. This can be connected to `MetricsSystem`, also can be collected back to driver as you did before.

Please help to review, your comments are greatly appreciated, thanks!

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: common/network-common/src/test/java/org/apache/spark/network/util/NettyMemoryMetricsSuite.java
Comment (dongjoon-hyun) [inline common/network-common/src/test/java/org/apache/spark/network/util/NettyMemoryMetricsSuite.java]: Indentation?

**Comment by jerryshao** [comment]
Thanks @dongjoon-hyun for your review, will change it soon.

**Comment by jerryshao** [comment]
@jiangxb1987 @zsxwing would you please help to review this PR, thanks!

**Comment by jiangxb1987** [comment]
I'll review this tomorrow.

**Comment by zsxwing** [comment]
Could you bump the netty version to use its new APIs rather than reflection?

**Comment by jerryshao** [comment]
Thanks @zsxwing . I was thinking to expose the details of memory allocation in Netty for user to monitor and tune, user could filter out unrelated metrics. Maybe you're right, it is too verbose and too detailed to expose each arena's detail, let me change the codes.

File: pom.xml
Comment (jerryshao) [inline pom.xml]: Upgrade Netty version 4.0.47 which has the public API for `ByteBufAllocatorMetric`, what do you think @zsxwing ?

File: common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java
Comment (jerryshao) [inline common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java]: Add a configuration to control whether to expose detailed metrics. Normally detailed per arena metrics will not be tracked, user could enable this through configuration, so that user has the ability to know the details of Netty memory allocation behavior, which could potentially guid to fine-grained tune.

**Comment by jerryshao** [comment]
Jenkins, retest this please.

**Comment by jerryshao** [comment]
Jenkins, retest this please.

File: common/network-common/src/main/java/org/apache/spark/network/util/NettyMemoryMetrics.java
Comment (squito) [inline common/network-common/src/main/java/org/apache/spark/network/util/NettyMemoryMetrics.java]: These metrics looks interesting, but do you think having them separated out by arena is helpful?  Seems like we want a view of these metrics which sums them over all arenas?

(Perhaps this is some ignornance of how arenas work in netty ... would help to seem some example metrics you get from it)

File: common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java
Comment (squito) [inline common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java]: it looks to me like this doesn't get exposed at all for the user to query -- eg. I dont' see it getting pulled into the MetricSystem.  is that happening somehow?  or this exposed another way?

**Review by squito — COMMENTED** [review]
Thanks for working on this @jerryshao I think this is a useful addition.  But I don't understand this part:

> This PR doesn't expose these metrics to any sink, to leverage this feature, still requires to connect to either MetricsSystem or collect them back to Driver to display.

Can you explain how this works to me?  Sorry if I am missing something, but it looks like these are not actually exposed at all.

File: common/network-common/src/main/java/org/apache/spark/network/util/NettyMemoryMetrics.java
Comment (squito) [inline common/network-common/src/main/java/org/apache/spark/network/util/NettyMemoryMetrics.java]: also just saw zsxwings comment:

> Does anyone really care about the details of PoolArenaMetric? Maybe just report https://github.com/netty/netty/blob/4.1/buffer/src/main/java/io/netty/buffer/ByteBufAllocatorMetric.java ?

that may be enough, rather than aggregating the details from the arenas.  again, an example of the info from these metrics would be helpful to decide if they're useful.

**Comment by jerryshao** [comment]
Hi @squito , thanks a lot for your review. This PR doesn't expose Netty metrics to MetricsSystem or collect to driver side (what did before), it just provides a way for `TransportClientFactory` and `TransportServer` to expose metrics to other components. So we can either pull this into MetricsSystem, or report to driver in the following PRs. That's why you don't find any way for user to query in this PR. This will be done it following ones.

**Comment by jerryshao** [comment]
@squito , here is the list of full metrics I dumped from running application. By default, if we don't enable verbose metric, only 2 metrics ("usedHeapMemory" and "usedDirectMemory") will be tracked. If we enable verbose metrics, then the details of each arena will be tracked, the number of detailed metrics is 16, so the total verbose metric number will be "16 * arena number * 2", here arena number is based on the number of cores, and "2" represents "direct arena" and "heap arena". 

I agree this might be too verbose for a user to track netty memory usage, so by default I will not enable verbose metric. Unless if developers wants to know the detailed behaviors of network memory usage, then they could enable this configuration, this might be useful for tuning and performance analysis. What do you think?

```
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.directArena0.numActiveAllocations
             value = 0
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.directArena0.numActiveBytes
             value = 16777216
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.directArena0.numActiveHugeAllocations
             value = 0
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.directArena0.numActiveNormalAllocations
             value = 0
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.directArena0.numActiveSmallAllocations
             value = 0
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.directArena0.numActiveTinyAllocations
             value = 0
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.directArena0.numAllocations
             value = 7
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.directArena0.numDeallocations
             value = 7
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.directArena0.numHugeAllocations
             value = 0
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.directArena0.numHugeDeallocations
             value = 0
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.directArena0.numNormalAllocations
             value = 2
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.directArena0.numNormalDeallocations
             value = 2
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.directArena0.numSmallAllocations
             value = 3
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.directArena0.numSmallDeallocations
             value = 3
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.directArena0.numTinyAllocations
             value = 2
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.directArena0.numTinyDeallocations
             value = 2
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.heapArena0.numActiveAllocations
             value = 0
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.heapArena0.numActiveBytes
             value = 16777216
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.heapArena0.numActiveHugeAllocations
             value = 0
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.heapArena0.numActiveNormalAllocations
             value = 0
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.heapArena0.numActiveSmallAllocations
             value = 0
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.heapArena0.numActiveTinyAllocations
             value = 0
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.heapArena0.numAllocations
             value = 4
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.heapArena0.numDeallocations
             value = 4
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.heapArena0.numHugeAllocations
             value = 0
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.heapArena0.numHugeDeallocations
             value = 0
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.heapArena0.numNormalAllocations
             value = 0
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.heapArena0.numNormalDeallocations
             value = 0
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.heapArena0.numSmallAllocations
             value = 0
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.heapArena0.numSmallDeallocations
             value = 0
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.heapArena0.numTinyAllocations
             value = 4
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.heapArena0.numTinyDeallocations
             value = 4
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.usedDirectMemory
             value = 16777216
application_1503394031686_0002.1.NettyBlockTransfer.shuffle-client.usedHeapMemory
             value = 16777216
```

**Comment by squito** [comment]
thanks for the added info @jerryshao.

(a) ok makes sense now about the use of this -- it is not exposed now, you plan to expose it in future changes.  That is fine, however I would like to at least see a *plan* for how you want to expose this before we merge this.  I think past efforts have gotten tripped on how we'd expose it in the driver; perhaps the simplest thing to do is expose it to the metric system?  There just isn't any point in this unless we're going to do something with it.

(b) thanks for including the example of the metrics.  But one minor thing I noticed -- are you at all surprised that the direct & heap memory are exactly the same?  Are things getting mixed up somewhere?  Or maybe those are just the values from the initialization netty always does.  I definitely think that is more metrics than most users care about, and would want verbose off by default.  I am wondering if those metrics are useful at all ... but I guess they are pretty cheap, and maybe there is some scenario where a user would care about some of them, eg. `numHugeAllocations`.

**Comment by jerryshao** [comment]
@squito , as the next following step, I would expose these metrics with MetricsSystem, I'm thinking of exposing shuffle related Netty memory usage. For RPC related memory usage, I'm not fully sure about the value of exposing them, though exposing them is a quite cheap call.

And furthermore, I think we can collect them in the driver side and display either through REST API or via web UI. But I think part of should be considered well on how to deliver to the end users.

As for the content of netty memory metrics, by default only two major memory usage will be exposed unless we enable verbose. I was thinking of picking some detailed metrics, but I found it is hard to decide which one is more important over others, so instead I chose to expose all of them and let user to decide which one is important. Also like you mentioned they're quite cheap and off by default, so I think it might be good to also optionally expose detailed infos.

> are you at all surprised that the direct & heap memory are exactly the same? Are things getting mixed up somewhere? Or maybe those are just the values from the initialization netty always does. 

I think it doesn't mix up anything. AFAIK it is what Netty `PooledByteBufAllocator` does at initialization, this 16MB is the default chunk size. I guess at initialization, Netty will create one chunk per arena.

**Comment by jerryshao** [comment]
@squito can you please review again? Thanks.

**Comment by squito** [comment]
lgtm

any more thoughts @jiangxb1987 @zsxwing  ?

**Comment by jiangxb1987** [comment]
The change itself looks good, but I'm afraid it's a bit too verbose to add all of those metrics. cc @zsxwing .

**Comment by zsxwing** [comment]
LGTM since enableVerboseMetrics is off by default. Merging to master. Thanks!

**Comment by zsxwing** [comment]
retest this please

**Comment by zsxwing** [comment]
Just realized the last build is 18 days ago. Triggered a new now. Will merge after passing tests.

**Review by jiangxb1987 — APPROVED** [review]
LGTM

**Comment by zsxwing** [comment]
Merging to master.

**Pull request closed without merging**
